### PR TITLE
Get rid of our custom mutex and condvar implementations too

### PR DESCRIPTION
**Pull request by hrydgard**
Also makes a ton of mutexes that used to be recursive, not. Learned that doing cond vars with recursive mutexes is quite dangerous (might be double locked when the notify comes, for example).

### CODE REVIEW

**Comment by hrydgard** [comment]
Not really, though it's good to have this done before doing that. The implementations I'm replacing are from back when std::mutex and std::condition_variable were not yet in the C++ standard, and now that there's a standard, our old code was only in the way. It even caused strange build problems and symbol collisions on my (not yet published) Windows UWP branch, so I felt it was time to clean house.

File: Common/ThreadSafeList.h [hunk 1]
```diff
@@ -26,60 +26,60 @@ class ThreadSafeList {
...
 		return list.pop_back();
 	}
 
 	inline void push_back(const T &v) {
-		lock_guard guard(lock);
+		std::lock_guard<std::mutex> guard(lock);
 		return list.push_back(v);
 	}
 
 	bool empty() const {
```
Comment (InusualZ) [inline Common/ThreadSafeList.h]: I have an question maybe not related to this PR but to the class itself. Why this method `empty` need an mutex. From my understanding you use an mutex if your accessed data is going to be modified?

Comment (hrydgard) [inline Common/ThreadSafeList.h]: It's probably not really necessary in this case, you're right. Though if multiple pieces of data might be modified elsewhere, and you want a "consistent" view of it, it still makes sense to lock while just reading.

**Comment by LunaMoo** [comment]
@hrydgard seems this change also broke android in some way, see http://forums.ppsspp.org/showthread.php?tid=22662 ~ black screen in UI before even running a game / on Galaxy S5.

File: Common/ThreadSafeList.h (hunk 1 above)
Comment (unknownbrackets) [inline Common/ThreadSafeList.h]: I'm not sure that `std::list<>` makes any guarantees about threadsafety.  For example, a thread switch may occur in the middle of a `list.clear()` which might result in `list.empty()` dereferencing a null pointer - just theoretically - in some STL implementation.

I don't think we're allowed to assume that even read-only methods are threadsafe in STL.

-[Unknown]

Comment (Orphis) [inline Common/ThreadSafeList.h]: I believe it depends on the implementation and they rarely make any make any guarantees. But the "const" operations should be mostly fine in different threads as they rarely mutate elements under the hood.

But clear isn't const, so you definitely need locking around it and any other access that might appear concurrently.

**Pull request merged**